[feature] Migration#remove_index

donaldong opened the pull request:
This adds a migration method remove_index.

### Test plan
- [x] Added a test case to verify this behavior.

donaldong commented:
rebase

donaldong commented:
improve test case

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/9f56836f1c67091de874201356e177803928b10f..eeb54d8959f55de6bd432884bb7850875094376d).

katyho commented on lib/redcord/migration/index.rb:
```diff
@@ -0,0 +1,24 @@
+    end
+
+    model.redis.srem("#{model.model_key}:#{attr_set_name}", index_name)
+    model.redis.del("#{model.model_key}:#{index_name}")
```
- For index attributes, wouldn't we want to scan through all values of the index attribute and delete it?
e.g. if `attr: a, String, index: true` is an attribute with values "a" and "b", then we'd have to delete the index sets `Redcord:Model:a:a` and `Redcord:Model:a:b` 

- I think this works for range index attributes; nil values for range index attributes are also specially stored as normal sets, so we would also want to delete `Redcord:Model:range_attr:`

Another potential concern is redis [del](https://redis.io/commands/del) time complexity is O(M) where M is the number of elements in the set. This could be potentially slow for the id set.

donaldong replied:
> * For index attributes, wouldn't we want to scan through all values of the index attribute and delete it?
> nil values for range index attributes are also specially stored as normal sets, so we would also want to delete Redcord:Model:range_attr:

Ah, you're absolutely right! will  update and add a test case for these.

> Another potential concern is redis [del](https://redis.io/commands/del) time complexity is O(M) where M is the number of elements in the set. This could be potentially slow for the id set.

good concern! Though it might be a necessary operation?

katyho replied:
What do you think about using spop? https://redis.io/commands/spop

donaldong replied:
I like it! feels like this could minimize the blocking caused by running migrations

donaldong commented:
address comments

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/032239afbee21885582a04711b54f6f05af49222..f0f8c849860625e558803eef290f5ecb3cf0742c).

donaldong commented:
update comment

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/69001125360f7745c9ac56dd7ca0d0a636d2fbe9..b994506c5d4f2900dc5a2be1f4a60879bce954f0).

donaldong commented:
update

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/30bc087fc59de6ede805ada0dc959005c9b1ce33..d13a689d7996f5c201410bd03fa5155d9dc85b60).

katyho commented on lib/redcord/migration/index.rb:
```diff
@@ -0,0 +1,66 @@
+        index_name: index_name,
+      )
+
+      model.redis.scan_each(match: "#{model.model_key}:#{index_name}:") { |key| _del_set(model, key) }
```
I think you can use `sscan_each` here to directly iterate over the set (since we know what the key is for the nil range index set)

donaldong replied:
well we still need to loop trough all the sets tho. given a set, we still need to call spop in sscan_each, so not entirely necessary i think.

katyho replied:
We know exactly what the key of the set is in this case, so we shouldn't have to loop through all the sets right? I think scan with match will still result in more iterations since it has to scan through all keys

https://redis.io/commands/scan#the-match-option

donaldong replied:
oooh right there's only one set here lol

katyho commented on lib/redcord/migration/index.rb:
```diff
@@ -0,0 +1,66 @@
+
+      model.redis.scan_each(match: "#{model.model_key}:#{index_name}:") { |key| _del_set(model, key) }
+      model.redis.scan_each(match: "#{model.model_key}:#{index_name}:") { |key| model.redis.del(key) }
+      model.redis.scan_each(match: "#{model.model_key}:#{index_name}") { |key| _del_zset(model, key) }
```
We can use `zscan_each` here as well (unless there's some inclination to use `scan` instead.

donaldong replied:
similar as above

donaldong commented on lib/redcord/migration/index.rb:
```diff
@@ -0,0 +1,66 @@
+
+  sig { params(model: T.class_of(Redcord::Base), index_name: Symbol).void }
+  def remove_index(model, index_name)
+    if model.redis.sismember("#{model.model_key}:index_attrs", index_name)
```
since we're adding the remove_index migration, the attributes have been changed on the ruby side; so we'd want to query from db instead

donaldong commented:
address comments

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/8664ad5180161ffdd031074acc5f78d18c67db88..3ddd1c0cb720f2c6e90213db3ff348fd76cfde5a).

donaldong commented:
update

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/3db81670597c47d716242124f60b750d8ec6198a..9509ec4ccde1422064a83b871897484b4164ab09).

katyho commented on lib/redcord/migration/index.rb:
```diff
@@ -0,0 +1,70 @@
+    # Use SPOP here to minimize blocking
+    loop do
+      break unless model.redis.spop(key)
+    end
```
Not sure if an empty set vs non-existent key matters, but should we also delete the key after we've deleted everything in the set?

katyho reviewed (approved):
🎉

donaldong commented:
oops, good catch!

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/df47882b89716e6901c39df25b78cca32efedb24..656c2748113bc0b00a348b7ac2aeaac3b19c9035).

donaldong commented:
zscan_each instead of zpopmin

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/2faaf59fbcd61472e394b44f96f9a9ccfb6b5caa..376ccf9fa781a7cf2ab6eceb4b953a3d2e79f5db).

katyho commented on lib/redcord/migration/index.rb:
```diff
@@ -0,0 +1,68 @@
+  sig { params(model: T.class_of(Redcord::Base), key: String).void }
+  def _del_zset(model, key)
+    # ZPOPMIN might not be avaliable on old redis servers
+    model.redis.zscan_each(match: key) do |id, _|
```
Ah nice catch to use zscan_each here! You can also remove the zpopmin comment, but otherwise this lgtm

donaldong replied:
actually the comment is intentional. ideally, we'd want to use zpopmin I think, but sadly its not always available
